### PR TITLE
Add Loopback_test File

### PR DESCRIPTION
**Pull request by naveedjamil**
Signed-off-by: Naveed Jamil <naveed.jamil@tenpearl.com>

**- What I did**
 increases the test coverage for the package pkg/loopback.
**- Why I did**
Initially there were no test files present in this module.
**- How I did it**
Include a flag for pkg/loopback in test-unit file.
**- How to verify it**
Run make test-unit
**- A picture of a cute animal (not mandatory but encouraged)**
Sorry

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -72,7 +72,8 @@ GIT_BRANCH_CLEAN := $(shell echo $(GIT_BRANCH) | sed -e "s/[^[:alnum:]]/-/g")
 DOCKER_IMAGE := docker-dev$(if $(GIT_BRANCH_CLEAN),:$(GIT_BRANCH_CLEAN))
 DOCKER_PORT_FORWARD := $(if $(DOCKER_PORT),-p "$(DOCKER_PORT)",)
 
-DOCKER_FLAGS := docker run --rm -i --privileged $(DOCKER_CONTAINER_NAME) $(DOCKER_ENVS) $(DOCKER_MOUNT) $(DOCKER_PORT_FORWARD)
+#TODO: This -v flag needs to be moved
+DOCKER_FLAGS := docker run -v /var/tmp:/var/tmp --rm -i --privileged $(DOCKER_CONTAINER_NAME) $(DOCKER_ENVS) $(DOCKER_MOUNT) $(DOCKER_PORT_FORWARD)
```
Comment (AkihiroSuda) [inline Makefile]: Let's just mount temps under `/tmp`.

File: hack/make/test-unit [hunk 2]
```diff
@@ -48,8 +48,8 @@ bundle_test_unit() {
 			| grep -v github.com/docker/docker/man \
 			| grep -v github.com/docker/docker/integration-cli)
 	fi
-
 	go test -cover -ldflags "$LDFLAGS" "${BUILDFLAGS[@]}" $TESTFLAGS $pkg_list
+	go test -cover -ldflags "$LDFLAGS" "${BUILDFLAGS[@]}" $TESTFLAGS github.com/docker/docker/pkg/loopback -test.root
```
Comment (AkihiroSuda) [inline hack/make/test-unit]: This is unneeded, `make unit-test` should execute the test as root

File: pkg/loopback/loopback_test.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	if !rootEnabled {
+		t.Skip("skipping test that requires root")
+		return
+	}
+	require.Equal(t, 0, os.Getuid(), "This test must be run as root.")
+}
+
+func createSparseFile() (string, error) {
+	size := int64(10 * 1024 * 1024)
+	fd, err := ioutil.TempFile(filePath, "sparseFile")
```
Comment (AkihiroSuda) [inline pkg/loopback/loopback_test.go]: You can use an empty string as `filePath` here

File: pkg/loopback/loopback_test.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+	require.Equal(t, loopDevice.Name(), confirmLoopDevice.Name())
+}
+
+func TestSetCapacity(t *testing.T) {
+	RequiresRoot(t)
+	fileName, err := createSparseFile()
+	require.NoError(t, err, "Error creating sparse file")
+	defer os.Remove(fileName)
+
+	loopDevice, err := os.Open("/dev/loop5")
```
Comment (AkihiroSuda) [inline pkg/loopback/loopback_test.go]: This seems to fail nondeterministically

File: Makefile (hunk 1 above)
Comment (naveed-jamil-tenpearls) [inline Makefile]: Ok. Changed the mounted path from `/var/tmp` to `/tmp`.

File: hack/make/test-unit (hunk 2 above)
Comment (naveed-jamil-tenpearls) [inline hack/make/test-unit]: Test cases of all the packages are executed with the same flag. So if we provide `-test.root` to earlier command we will receive and error `flag provided but not defined: test.root` for all of the packages and we will have to define this flag in all packages. Which is why it was decided to be put explicitly for this package only.

File: Makefile (hunk 1 above)
Comment (AkihiroSuda) [inline Makefile]: I meant putting `--mount type=tmpfs,target=/tmp` to `$(DOCKER_MOUNT)` if you need your temporary directory to be tmpfs. (but why do you need that?)

File: pkg/loopback/loopback_test.go (hunk 3 above)
Comment (naveed-jamil-tenpearls) [inline pkg/loopback/loopback_test.go]: We need to provide a filepath which is the mounted volume inside the docker container, otherwise the function `file.stat()` provides different values for device and inode. To bypass this we had to mount a folder and create a file in that so that we can get accurate system level values of device and inode.

File: hack/make/test-unit (hunk 2 above)
Comment (AkihiroSuda) [inline hack/make/test-unit]: Other tests should have been already failing under non-root, but if you want to skip your test on non-root, I suggest checking UID as in https://github.com/moby/moby/blob/master/pkg/mount/mounter_linux_test.go#L14

File: Makefile (hunk 1 above)
Comment (naveed-jamil-tenpearls) [inline Makefile]: No I don't specifically need the temporary directory as such, but I need to mount a folder to create a sparse file in it. I have explained why i need that in my other comment.

File: hack/make/test-unit (hunk 2 above)
Comment (naveed-jamil-tenpearls) [inline hack/make/test-unit]: Yes I have already done that as you can see in the test file in the function `RequiresRoot(t *testing.T)`. If you provide the flag and there is no root access the test case fails. You have to provide both the flag and root access for these test cases to work.

Comment (aaronlehmann) [inline hack/make/test-unit]: @AkihiroSuda: I suggested adding this flag because I feel that `go test <packagename>` without extra arguments should work without expecting a certain environment or level of privilege. It would be unexpected for someone who downloads this package to run `go test` and have the tests fail because `go test` was not run as root, or, worse, to have the test mess with loop devices on their system.

The "requires root" flag is the way containerd's tests deal with this: https://github.com/containerd/containerd/blob/master/testutil/helpers.go

I realize there are existing tests that assume root privileges and interact with system resources, but rather than adding to the problem I think we should migrate those tests to use this new flag.

File: pkg/loopback/loopback_test.go (hunk 4 above)
Comment (naveed-jamil-tenpearls) [inline pkg/loopback/loopback_test.go]: Refactored the code to have a more thorough check.

File: pkg/loopback/loopback_test.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const filePath = "/tmp/"
+
+var rootEnabled bool
+
+func init() {
+	flag.BoolVar(&rootEnabled, "test.root", false, "enable tests that require root")
```
Comment (dnephin) [inline pkg/loopback/loopback_test.go]: Same comment as https://github.com/moby/moby/pull/33520/files#r125745827, I don't think we should use a flag. This should just skip if `os.Getuid() != 0`.

File: Makefile [hunk 6]
```diff
@@ -72,7 +72,8 @@ GIT_BRANCH_CLEAN := $(shell echo $(GIT_BRANCH) | sed -e "s/[^[:alnum:]]/-/g")
 DOCKER_IMAGE := docker-dev$(if $(GIT_BRANCH_CLEAN),:$(GIT_BRANCH_CLEAN))
 DOCKER_PORT_FORWARD := $(if $(DOCKER_PORT),-p "$(DOCKER_PORT)",)
 
-DOCKER_FLAGS := docker run --rm -i --privileged $(DOCKER_CONTAINER_NAME) $(DOCKER_ENVS) $(DOCKER_MOUNT) $(DOCKER_PORT_FORWARD)
+#TODO: This -v flag needs to be moved
+DOCKER_FLAGS := docker run -v /var/tmp:/tmp --rm -i --privileged $(DOCKER_CONTAINER_NAME) $(DOCKER_ENVS) $(DOCKER_MOUNT) $(DOCKER_PORT_FORWARD)
```
Comment (dnephin) [inline Makefile]: Why is this necessary?

Could we create this path in the container instead?

**Review by dnephin — CHANGES_REQUESTED** [review]

File: pkg/loopback/loopback_test.go (hunk 5 above)
Comment (stevvooe) [inline pkg/loopback/loopback_test.go]: This is the right approach in Go. Relying on the guid isn't faulty, especially when it may make destructive changes.

Comment (dnephin) [inline pkg/loopback/loopback_test.go]: We still rely on the uid. There's an assert right after this!  At least in the `pkg/term` case I don't think there was anything destructive. What is destructive here?  And wouldn't skipping the tests with you don't have permissions be safer? When would you have uid == 0 and be the wrong user?

If we absolutely need a switch to turn this on or off (which I'm not convinced we do), we should use an environment variable.

**Comment by stevvooe** [comment]
@naveed-jamil-tenpearls Thanks for the submission!

My first reaction is looking at the reasoning behind this change. I would ask the following questions:

Why does this need to be tested?
Is it a common source of error?
Have we need a large number of defects related to loopback devices?
Is this an exported Go interface we want to support?

The last question is most important and I think the answer is no. This was a part of the devicemapper package before being split out. The exported interface isn't generally useful outside of that use case and I don't think we should enshrine it with unit tests.

Arguably, this should just be merged back into the devicemapper package. This was probably tested as part of devicemapper before it was split out.

Unless you intend to use for something, which is completely valid, it might be best to merge this back into devicemapper and ensure that is tested heavily.

File: Makefile (hunk 6 above)
Comment (naveed-jamil-tenpearls) [inline Makefile]: This is necessary because we need to provide a file path which is present in a mounted volume inside the docker container, otherwise the function file.stat() provides different values for device and inode. This way we get the accurate system level values for device and inode of the given file.

File: pkg/loopback/loopback_test.go (hunk 5 above)
Comment (naveed-jamil-tenpearls) [inline pkg/loopback/loopback_test.go]: @dnephin I agree with you. But, similar flag was put in this [PR](https://github.com/moby/moby/pull/33520) which was merged. So I would suggest for consistency we should change this for the already merged PR as well. 
So which approach should we continue with ?

Comment (dnephin) [inline pkg/loopback/loopback_test.go]: I have a PR open to remove the other flag (and make some other changes) #33987

File: Makefile (hunk 6 above)
Comment (dnephin) [inline Makefile]: Can you point me at which test lines fail without this?

File: pkg/loopback/loopback_test.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+	defer os.Remove(fileName)
+
+	fd, err := os.Open(fileName)
+	require.NoError(t, err, "Error opening file")
+	defer fd.Close()
+
+	loopDevice, err := AttachLoopDevice(fileName)
+	require.NoError(t, err, "Error attaching loop device")
+
+	confirmLoopDevice := FindLoopDeviceFor(fd)
```
Comment (naveed-jamil-tenpearls) [inline pkg/loopback/loopback_test.go]: @dnephin the function `FindLoopDeviceFor` returns `nil` if we don't create a file inside a mounted volume. Also the targetINode and targetDevice present in line 77 and 78 respectively return different values while the file is not present in a mounted directory, hence the test fails. So for this we have to mount a volume while executing docker run.

**Comment by naveed-jamil-tenpearls** [comment]
@stevvooe sorry for the late reply. I must've missed your comment.
Well, the reason I wrote these tests was that the packages present in pkg module are stand alone modules which are capable of being run separately. So the primary focus was to increase the coverage of this package so that any other module which wants to use this package can extract this module from moby/moby and use it there. These tests will be specific to this package only and can run there separately. Whilst I agree that pkg/loopback is being covered in devicemapper and that we can move these tests there, I believe that because of the nature of the packages present in pkg, we should have coverage specific to these packages. So that they can be conveniently used elsewhere.

**Comment by dnephin** [comment]
Thank you for the additional information.  It looks like the file needs to exist on a loopback device. Is there any reason we can't create that during the test?

Re: testing. The plan in #32989 suggests this package should be moved to `devmapper/loopback` (https://gist.github.com/dnephin/35dc10f6b6b7017f058a71908b301d38#file-_-proposal-L92-L93) and shouldn't remain a standalone package. 

I think that would be a better idea. Some of these functions are not written in a way that allows us to write reasonable unit tests for them. They are probably better offf as implementation details of the devmapper package.

File: pkg/loopback/loopback_test.go (hunk 5 above)
Comment (stevvooe) [inline pkg/loopback/loopback_test.go]: @dnephin The problem is running a destructive test as `uid=0` and not expecting to have the test run whatever it did. This makes root-level tests opt-in, rather than opt out.

Let's not assume that everyone is using moby's build system to build everything and that it is widely safe. As it is, they take way to long to run, so assuming that this will always be safe to run as uid 0 is a poor assumption.

Comment (dnephin) [inline pkg/loopback/loopback_test.go]: I agree that if a test is potentially destructive it should be opt in. I'm still not clear about what exactly is destructive or not safe about these tests, but let's say that they are.

There are other options for opt-in that I think are better suited for a test suite. Build tags are good if the entire file should be ignored, and environment variables are great for individual tests.

I think the advantage to these options is that you don't have to special case testing of any package. You can set the build tag or environment variable for everything, and not  worry about which packages happen to be selected by the user on a specific test run.

> As it is, they take way to long to run

Yes they do, but the unit tests are not really a significant part of that time. Most of the time is running way too many integration tests.

**Comment by naveed-jamil-tenpearls** [comment]
I see, so I am closing this PR.

File: pkg/loopback/loopback_test.go (hunk 5 above)
Comment (stevvooe) [inline pkg/loopback/loopback_test.go]: @dnephin It is not about whether a particular test is destructive. It is about the potential for destruction as a root-level test.

Test flags a proven pattern in Go that work much better than environment variables or build flags. Build flags often make it hard to include sets of tests and environment variables have problems with documentation. Test flags are easy to build and add and are self-documenting, since they show in the help of a test.

Comment (dnephin) [inline pkg/loopback/loopback_test.go]: > It is about the potential for destruction as a root-level test.

What is a "root-level test" ? Do you mean a test that requires root to run? If a user is running tests as root, they are already assuming that risk. 

> Test flags are easy to build and add and are self-documenting, since they show in the help of a test.

Do you really think that running `--help` on a test binary is a common use case? The default `go test` doesn't even expose the binary.

Comment (stevvooe) [inline pkg/loopback/loopback_test.go]: @dnephin Let's please follow the Go best practices rather than reinventing the wheel.

```
-httptest.serve string
    	if non-empty, httptest.NewServer serves on this address and blocks
-test.root
    	enable tests that require root
```
